File: mmoTest/Assets/Script/Data/Proto/WorldMap/WorldMap_OtherRoleLeaveProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2018-02-25 22:40:38
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 服务器广播其他角色离开场景消息
/// </summary>
public struct WorldMap_OtherRoleLeaveProto : IProto
{
    public ushort ProtoCode { get { return 13006; } }

    public int RoleId; //角色编号

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(RoleId);
            return ms.ToArray();
        }
    }

    public static WorldMap_OtherRoleLeaveProto GetProto(byte[] buffer)
    {
        WorldMap_OtherRoleLeaveProto proto = new WorldMap_OtherRoleLeaveProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.RoleId = ms.ReadInt();
        }
        return proto;
    }
}

File: MMORPGServer/Proto/CreateLuaProto/WorldMap_OtherRoleLeaveProto.lua
--服务器广播其他角色离开场景消息
WorldMap_OtherRoleLeaveProto = { ProtoCode = 13006, RoleId = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
WorldMap_OtherRoleLeaveProto.__index = WorldMap_OtherRoleLeaveProto;

function WorldMap_OtherRoleLeaveProto.New()
    local self = { }; --初始化self
    setmetatable(self, WorldMap_OtherRoleLeaveProto); --将self的元表设定为Class
    return self;
end


--发送协议
function WorldMap_OtherRoleLeaveProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.RoleId);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function WorldMap_OtherRoleLeaveProto.GetProto(buffer)

    local proto = WorldMap_OtherRoleLeaveProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.RoleId = ms:ReadInt();

    ms:Dispose();
    return proto;
end